File: analytics/metrics-platform/src/test/java/org/wikimedia/metrics_platform/SessionControllerTest.java
package org.wikimedia.metrics_platform;

import static java.time.temporal.ChronoUnit.HOURS;
import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;

import org.junit.jupiter.api.Test;

class SessionControllerTest {

    @Test void testSessionExpiry() {
        Instant oneHourAgo = Instant.now().minus(1, HOURS);
        SessionController sessionController = new SessionController(oneHourAgo);
        assertThat(sessionController.sessionExpired()).isTrue();
    }

    @Test void testTouchSession() {
        Instant oneHourAgo = Instant.now().minus(1, HOURS);
        SessionController sessionController = new SessionController(oneHourAgo);
        String sessionId1 = sessionController.getSessionId();
        sessionController.touchSession();
        assertThat(sessionController.sessionExpired()).isFalse();

        String sessionId2 = sessionController.getSessionId();

        assertThat(sessionId1).isNotEqualTo(sessionId2);
    }

    @Test void testSessionIdLength() {
        Instant twoHoursAgo = Instant.now().minus(2, HOURS);
        SessionController sessionController = new SessionController(twoHoursAgo);
        String sessionId = sessionController.getSessionId();
        assertThat(sessionId.length()).isEqualTo(20);
    }
}

File: analytics/testkitchen/src/test/java/org/wikimedia/testkitchen/SessionControllerTest.kt
package org.wikimedia.testkitchen

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant
import java.time.temporal.ChronoUnit

internal class SessionControllerTest {
    @Test
    fun testSessionExpiry() {
        val oneHourAgo = Instant.now().minus(1, ChronoUnit.HOURS)
        val sessionController = SessionController(oneHourAgo)
        assertTrue(sessionController.sessionExpired())
    }

    @Test
    fun testTouchSession() {
        val oneHourAgo = Instant.now().minus(1, ChronoUnit.HOURS)
        val sessionController = SessionController(oneHourAgo)
        val sessionId1 = sessionController.sessionId
        sessionController.touchSession()
        assertFalse(sessionController.sessionExpired())

        val sessionId2 = sessionController.sessionId
        assertNotEquals(sessionId1, sessionId2)
    }

    @Test
    fun testSessionIdLength() {
        val twoHoursAgo = java.time.Instant.now().minus(2, java.time.temporal.ChronoUnit.HOURS)
        val sessionController = SessionController(twoHoursAgo)
        val sessionId = sessionController.sessionId
        assertEquals(20, sessionId.length)
    }
}
